grpc/authz/authz.go

```diff
@@ -29,6 +30,87 @@ type Base struct {
 	ValidateScopeFunc         func(tokenClaims []byte, scopes []string) error
 	Audience                  string
 	DisableAudienceValidation bool
+	// HSKey for use with HS algorithms
+	HSKey []byte
+	// RSAPublicKey for use with RS algorithms
+	RSAPublicKey *rsa.PublicKey
+}
+
+// ConfigOption for setting optional values
+type ConfigOption func(*Config)
+
+// WithScopeMapping sets the ScopeMapping option
+func WithScopeMapping(scopeMap map[string][]string) ConfigOption {
+	return func(args *Config) { args.ScopeMapping = scopeMap }
+}
+
+// WithValidateScopeFunc sets the ValidateScopeFunc option
+func WithValidateScopeFunc(scopeFunc func(tokenClaims []byte, scopes []string) error) ConfigOption {
+	return func(args *Config) { args.ValidateScopeFunc = scopeFunc }
+}
+
+// WithAudience sets the audience
+func WithAudience(aud string) ConfigOption {
+	return func(args *Config) { args.Audience = aud }
+}
+
+// WithDisableAudienceValidation sets the WithDisableAudienceValidation option
+func WithDisableAudienceValidation(disable bool) ConfigOption {
+	return func(args *Config) { args.DisableAudienceValidation = disable }
+}
+
+// WithHSKey sets the HS key
+func WithHSKey(hKey []byte) ConfigOption {
+	return func(args *Config) { args.HSKey = hKey }
+}
+
+// WithRSAPubKey sets the RSA public key
+func WithRSAPubKey(rsaPubKey *rsa.PublicKey) ConfigOption {
+	return func(args *Config) { args.RSAPublicKey = rsaPubKey }
+}
+
+// NewConfig returns a new config with options
+func NewConfig(algo jwt.Algorithm, opts ...ConfigOption) *Config {
+	defaultConfig := &Config{
+		Algorithm:         algo,
+		ValidateScopeFunc: func(tokenClaims []byte, scopes []string) error { return nil },
+	}
+	for _, opt := range opts {
+		opt(defaultConfig)
+	}
+	return defaultConfig
+}
+
+// AuthFunc authorization function
+func (c *Config) AuthFunc(ctx context.Context) (context.Context, error) {
+	token, protected, err := c.doProtected(ctx)
+	if err != nil {
+		return ctx, err
+	}
+	if !protected {
+		return ctx, nil
+	}
+	var verifier jwt.Verifier
+	switch c.Algorithm {
+	case jwt.HS256, jwt.HS384, jwt.HS512:
+		verifier, err = jwt.NewVerifierHS(c.Algorithm, c.HSKey)
+		if err != nil {
+			return ctx, status.Errorf(codes.FailedPrecondition, "verifier error: %v", err.Error())
+		}
+	case jwt.RS256, jwt.RS384, jwt.RS512:
+		verifier, err = jwt.NewVerifierRS(c.Algorithm, c.RSAPublicKey)
+		if err != nil {
+			return ctx, status.Errorf(codes.FailedPrecondition, "verifier error: %v", err.Error())
+		}
+	default:
+		return ctx, status.Errorf(codes.Unimplemented, "algorithm is not supported: %T", c.Algorithm)
+	}
```

Comment:
This should also be done in NewConfig to catch errors earlier.

Reply:
Hmm....the side effects of checking in `NewClient` might not be worth it. If we use the same logic to move this check into `NewConfig` we probably should also move a validation check (currently done in `AuthFunc` when calling `NewVerifier`) against the provided key to make sure it matches the specified algorithm. not sure we want that.